### PR TITLE
Dijkstra arbitrary weights

### PR DESCRIPTION
**Pull request by EliasLF**
## Description

Replacing fixed edge costs in Dijkstra implementation with an edge weight matrix. Thereby enabling future expansions (fidelity-aware distances, semi-directional architectures, etc.)

## Checklist:

- [x] The pull request only contains commits that are related to it.
- [x] I have added appropriate tests and documentation.
- [ ] I have made sure that all CI jobs on GitHub pass.
- [x] The pull request introduces no new warnings and follows the project's style guidelines.

### CODE REVIEW

File: include/Architecture.hpp [hunk 1]
```diff
@@ -364,21 +364,18 @@ class Architecture {
   void createDistanceTable();
   void createFidelityTable();
 
-  static double costHeuristicBidirectional(const Dijkstra::Node& node) {
-    auto length = node.cost - 1;
+  static double dijkstraNodeToCostNonFidelity(const Dijkstra::Node& node) {
```
Comment (burgholzer) [inline include/Architecture.hpp:367]: Just a very minor nit: the naming of this function is kind of non-ideal as there is no "fidelity" version. Might as well just call this function `dijkstraNodeToCost`.

File: test/test_general.cpp [hunk 2]
```diff
@@ -52,3 +53,65 @@ TEST(General, TestLineParsing) {
...
+  /*
+            (6)
+        .----------.
+       \/          \/
+  0 <-> 1 --> 2 <-> 3
+    (1)   (2)   (3)
+  */
+  CouplingMap cm = {{0, 1}, {1, 0}, {1, 2}, {2, 3}, {3, 2}, {1, 3}, {3, 1}};
+  Matrix      edgeWeights = {
+      {0, 1, 0, 0}, {1, 0, 2, 6}, {0, 15, 0, 3}, {0, 6, 3, 0}};
```
Comment (burgholzer) [inline test/test_general.cpp:67]: Is the `15` in the third entry here on purpose? Maybe I just interpreted the comment above in the wrong way, but to me it matches every entry but this `15`.

File: test/test_general.cpp [hunk 3]
```diff
@@ -52,3 +53,65 @@ TEST(General, TestLineParsing) {
...
+  EXPECT_EQ(distanceTable[2].size(), 4);
+  EXPECT_EQ(distanceTable[2][0], 10);
+  EXPECT_EQ(distanceTable[2][1], 9);
+  EXPECT_EQ(distanceTable[2][2], 0);
+  EXPECT_EQ(distanceTable[2][3], 3);
+  EXPECT_EQ(distanceTable[3].size(), 4);
+  EXPECT_EQ(distanceTable[3][0], 7);
+  EXPECT_EQ(distanceTable[3][1], 6);
+  EXPECT_EQ(distanceTable[3][2], 3);
+  EXPECT_EQ(distanceTable[3][3], 0);
```
Comment (burgholzer) [inline test/test_general.cpp:91]: I am not quite sure whether googletest is capable of that, but couldn't you write this in a much more compact fashion by specifying a `targetTable` (just like the edge weights matrix) and do `EXPECT_EQ(distanceTable, targetTable);`?

File: test/test_general.cpp (hunk 2 above)
Comment (burgholzer) [inline test/test_general.cpp:67]: Both of these could be declared `const`.

File: test/test_general.cpp [hunk 4]
```diff
@@ -52,3 +53,65 @@ TEST(General, TestLineParsing) {
...
+TEST(General, Dijkstra) {
+  /*
+            (6)
+        .----------.
+       \/          \/
+  0 <-> 1 --> 2 <-> 3
+    (1)   (2)   (3)
+  */
+  CouplingMap cm = {{0, 1}, {1, 0}, {1, 2}, {2, 3}, {3, 2}, {1, 3}, {3, 1}};
+  Matrix      edgeWeights = {
```
Comment (burgholzer) [inline test/test_general.cpp:66]: I think you can get around the clang-format error from the linter by inserting a newline in front of this line and reformatting. Always strange how two independent systems (clang-tidy locally, as well as pre-commit) say everything is alright, but a this system says no..

**Review by burgholzer — APPROVED** [review]
Many thanks. This looks good to me. I just have a couple of minor suggestions and small questions. Feel free to discard them. I also added a comment on how you might be able to circumvent the cpp-linter warning.

File: include/Architecture.hpp (hunk 1 above)
Comment (EliasLF) [reply include/Architecture.hpp:367]: I am going to submit another PR adding some preparatory functionality for fidelity-aware mapping (including a fidelity-aware version of this method) probably later this day. Since I did not know, if at that point this PR will already be merged, I just wanted to prevent merging conflicts with this naming. But if you want I can also rename it `dijkstraNodeToCost` for now.

File: test/test_general.cpp (hunk 2 above)
Comment (EliasLF) [reply test/test_general.cpp:67]: Good catch. Thank you for the hint. But this is in fact on purpose. Our Dijkstra implementation generally ignores directionality of edges. The edge weight of `15` (which is higher than the sum of all other weights combined) kind of simulates a unidirectional edge. (Just to make sure that Dijkstra takes the correct path in that edge case)

Comment (EliasLF) [reply test/test_general.cpp:67]: Right, thank you

File: include/Architecture.hpp (hunk 1 above)
Comment (burgholzer) [reply include/Architecture.hpp:367]: I thought that this might be the case. I think I still have a slight preference for just removing the "NonFidelity" from the function name as it doesn't really add to the description of what the function does, whereas the fidelity version clearly states that it does something similar to the function here, but also considering fidelity.
However, this is definitely not a strong opinion. So feel free to leave the naming as it is.

File: test/test_general.cpp (hunk 2 above)
Comment (burgholzer) [reply test/test_general.cpp:67]: That makes sense. Thanks for the explanation!

File: include/Architecture.hpp (hunk 1 above)
Comment (EliasLF) [reply include/Architecture.hpp]: Oh, I see, that is true. Alright, I changed it now.

File: test/test_general.cpp (hunk 3 above)
Comment (EliasLF) [reply test/test_general.cpp]: I did not think that googletest supports deep collection comparisons, but it actually works. Thank you for the suggestion. The code is much cleaner now that way.

File: test/test_general.cpp (hunk 4 above)
Comment (EliasLF) [reply test/test_general.cpp]: That worked, thank you! I have been slowly running out of ideas how to appease the linter there

**Review by burgholzer — APPROVED** [review]
LGTM 👍🏻 I'll merge once CI is satisfied.

File: test/test_general.cpp (hunk 4 above)
Comment (burgholzer) [reply test/test_general.cpp]: Yeah. Sometimes it's a pain.. over time you develop a repertoire of tricks to fix these kind of issues 😌
